src/Microsoft.Framework.OptionsModel/ConfigurationBinder.cs

```diff
@@ -0,0 +1,241 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Framework.ConfigurationModel;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.Framework.OptionsModel
+{
+    public static class ConfigurationBinder
+    {
+        public static TOptions Bind<TOptions>(IConfiguration configuration) where TOptions : new()
+        {
+            var model = new TOptions();
+            Bind(model, configuration);
+            return model;
+        }
+
+        public static void Bind(object model, IConfiguration configuration)
+        {
+            if (model == null)
+            {
+                return;
+            }
+
+            BindObjectProperties(model, configuration);
+        }
+
+        private static void BindObjectProperties(object obj, IConfiguration configuration)
+        {
+            foreach (var property in GetAllProperties(obj.GetType().GetTypeInfo()))
+            {
+                BindProperty(property, obj, configuration);
+            }
+        }
+
+        private static void BindProperty(PropertyInfo property, object propertyOwner, IConfiguration configuration)
+        {
+            configuration = configuration.GetSubKey(property.Name);
+
+            if (property.GetMethod == null || !property.GetMethod.IsPublic)
+            {
+                // We don't support set only properties
+                return;
+            }
+
+            var propertyValue = property.GetValue(propertyOwner);
+            var hasPublicSetter = property.SetMethod != null && property.SetMethod.IsPublic;
+
+            if (propertyValue == null && !hasPublicSetter)
+            {
+                // Property doesn't have a value and we cannot set it so there is no
+                // point in going further down the graph
+                return;
+            }
+
+            propertyValue = BindType(property.PropertyType, propertyValue, configuration);
+
+            if (propertyValue != null && hasPublicSetter)
+            {
+                property.SetValue(propertyOwner, propertyValue);
+            }
+        }
+
+
+        private static object BindType(Type type, object typeInstance, IConfiguration configuration)
+        {
+            var configValue = configuration.Get(null);
```

Comment:
Ah ok, so that's how you extract the value from config leaf nodes...  That syntax just looks odd to me.  There's no other sugar like configuration.Value?